src/internet_identity/tests/integration/v2_api/identity_register/dynamic_captcha.rs

```diff
@@ -0,0 +1,68 @@
+use crate::v2_api::authn_method_test_helpers::{
+    create_identity_with_authn_method, test_authn_method,
+};
+use canister_tests::api::internet_identity::api_v2;
+use canister_tests::framework::{
+    arg_with_dynamic_captcha, env, install_ii_canister_with_arg, test_principal, II_WASM,
+};
+use internet_identity_interface::internet_identity::types::RegistrationFlowNextStep;
+use std::time::Duration;
+
+#[test]
+fn should_not_require_captcha_below_threshold_rate() {
+    let env = env();
+    let canister_id =
+        install_ii_canister_with_arg(&env, II_WASM.clone(), arg_with_dynamic_captcha());
+    let authn_method = test_authn_method();
+
+    let flow_principal = test_principal(0);
+    let result = api_v2::identity_registration_start(&env, canister_id, flow_principal)
+        .expect("API call failed")
+        .expect("registration start failed");
+
+    assert!(matches!(result.next_step, RegistrationFlowNextStep::Finish));
+
+    api_v2::identity_registration_finish(&env, canister_id, flow_principal, &authn_method)
+        .expect("API call failed")
+        .expect("registration finish failed");
+}
+
+#[test]
+fn should_require_captcha_above_threshold_rate() {
+    let env = env();
+    let canister_id =
+        install_ii_canister_with_arg(&env, II_WASM.clone(), arg_with_dynamic_captcha());
+    let authn_method = test_authn_method();
+
+    // initialize a base rate of one registration every 2 seconds
+    for _ in 0..10 {
+        create_identity_with_authn_method(&env, canister_id, &authn_method);
+        env.advance_time(Duration::from_secs(2))
+    }
+
+    // Double the rate of registrations to one per second
+    // The 20% threshold rate should allow 5 registrations before the captcha kicks in
+    let flow_principal = test_principal(0);
```

Comment:
It doesn't matter that the principal for the registration is the same?

Reply:
If all previous flows are completed, then it does not matter (because flows are cleaned up on `finish`). But you're right, it's cleaner to use different ones.